### PR TITLE
Users/yimengyuan/privatednsrefactored

### PR DESCRIPTION
**Pull request by Yimyames**
# Change
- Refactored module to child resource structure
- Updated parameter file
- Updated documentation

Pipeline reference
[Network: Privatednszones](https://github.com/Yimyames/ResourceModules/actions/workflows/ms.network.privatednszones.yml)

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  location: location
+  tags: tags
+  properties: {
+    registrationEnabled: registrationEnabled
+    virtualNetwork: {
+      id: virtualNetworkId
+    }
+  }
+  parent: privateDnsZone
+}
```
Comment (MariusStorhaug) [inline arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep:42]: Check this with Alex. Not sure we have decided on this being the structure we want to go with yet. Before that decision is made, I would suggest that we keep it to the current way of doing this. @MrMCake What's your take on this? Should we not stay with the current agreement, while having the 2 module that POC the other approaches leaving the rest in line?

Current way of implementing this: <https://github.com/Azure/ResourceModules/blob/main/arm/Microsoft.Compute/virtualMachines/extensions/deploy.bicep#L47>

Issue describing the change, where no decision has been made, before we do the restructuring of modules:
https://github.com/Azure/ResourceModules/issues/480

**Review by MariusStorhaug — CHANGES_REQUESTED** [review]
Looking really good, seeing this is the first contribution!

File: arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep (hunk 1 above)
Comment (MrMCake) [reply arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep:42]: Hey @MariusStorhaug, should be fine here. We used the same in for example AutomationAccount. On this level it does not require any special treatment. Hence I'd keep it here

File: arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json [hunk 2]
```diff
@@ -2,18 +2,26 @@
...
+        "name": {
+            "value": "<<namePrefix>>.az-privdns-x-001"
         },
         "roleAssignments": {
             "value": [
                 {
                     "roleDefinitionIdOrName": "Reader",
                     "principalIds": [
-                        "cf33fea8-b30f-424f-ab73-c48d99e0b222"
+                        "e58511af-4da2-449c-a5cd-6a10271cfb83"
```
[the remark this answers is not in the document]
Comment (MrMCake) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json:13]: Good catch. We also use them in EventHub by the way. @Yimyames can you update the Id as described by @MariusStorhaug and re-run the pipeline?

[the remark this answers is not in the document]
Comment (Yimyames) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: Hi @MariusStorhaug, thanks for the review. I tested the token, it does not work yet - the ID works though.

[the remark this answers is not in the document]
Comment (MariusStorhaug) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: Did you integrate latest changes from upstream main? I see that it should be available from the upstream:main branch:
https://github.com/Azure/ResourceModules/blob/main/.github/actions/templates/deployModule/action.yml#L63

[the remark this answers is not in the document]
Comment (MariusStorhaug) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: Could also be that the variable needs to be added to your own fork/environment.
Docs on this seems to be somewhere here:
https://github.com/Azure/ResourceModules/blob/main/docs/wiki/GettingStarted.md#L213

File: arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep (hunk 1 above)
Comment (MariusStorhaug) [reply arm/Microsoft.Network/privateDnsZones/virtualNetworkLinks/deploy.bicep:42]: Ok!

File: arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (MariusStorhaug) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: Could also be that the variable needs to be added to your own fork/environment.
Docs on this seems to be somewhere here:
https://github.com/Azure/ResourceModules/blob/main/docs/wiki/GettingStarted.md#L213

[the remark this answers is not in the document]
Comment (MrMCake) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: Hey @MariusStorhaug , as this [PR](https://github.com/Azure/ResourceModules/pull/500) isn't merged yet, this won't work. @Yimyames tested it too and it did indeed fail

[the remark this answers is not in the document]
Comment (MariusStorhaug) [reply arm/Microsoft.Network/privateDnsZones/.parameters/parameters.json]: right, so that needs to come later. Agreed

**Review by MariusStorhaug — APPROVED** [review]
All of my comments are resolved. Approved

**Pull request merged**